time/mock/time.go

```diff
@@ -35,5 +83,46 @@ func (m *MockTime) SetAutoIncrement(t time.Duration) {
 }
 
 func (m *MockTime) IncrementTime(t time.Duration) {
+	m.lock.Lock()
+	defer m.lock.Unlock()
+
+	m.incrementTimeLockHeld(t)
+}
+
+func (m *MockTime) HasTimers() bool {
+	m.lock.Lock()
+	defer m.lock.Unlock()
+
+	return len(m.timers) > 0
+}
+
+func (m *MockTime) incrementTimeLockHeld(t time.Duration) {
+	if t == 0 {
+		return
+	}
+
 	m.currentTime = m.currentTime.Add(t)
+	logrus.WithField("increment", t).WithField("t", m.currentTime.Sub(startTime)).Info("Incrementing time")
+
+	if len(m.timers) == 0 {
+		return
+	}
+
+	sort.Slice(m.timers, func(i, j int) bool {
+		return m.timers[i].TimeToFire.Before(m.timers[j].TimeToFire)
+	})
+
+	logrus.WithField("delay", m.timers[0].TimeToFire.Sub(m.currentTime)).Info("Next timer.")
+
+	for len(m.timers) > 0 &&
+		(m.timers[0].TimeToFire.Before(m.currentTime) ||
+			m.timers[0].TimeToFire.Equal(m.currentTime)) {
+		logrus.WithField("timer", m.timers[0]).Info("Firing timer.")
+		select {
+		case m.timers[0].C <- m.currentTime: // Should never block since there channel has cap 1.
```

Comment:
`m.currentTime` is technically wrong, as it should be when the timer was due, not the (later) time now.  But I don't know if any timer consumers care about the difference.

Reply:
Even more technically, the timer sends the "current time" https://golang.org/pkg/time/#After on the channel, which could be after the desired fire time if the system was under load.

Agree that it'd probably be a better simulation if it sent the timer's TimeToFire.